presto-orc/src/main/java/com/facebook/presto/orc/reader/MapFlatSelectiveStreamReader.java

```diff
@@ -0,0 +1,830 @@
+        }
+
+        int streamPosition = 0;
+
+        int[] nonNullPositions = new int[positionCount];
+        int nonNullPositionCount = 0;
+        int nonNullSkipped = 0;
+
+        for (int i = 0; i < positionCount; i++) {
+            int position = positions[i];
+            if (position > streamPosition) {
+                nonNullSkipped += presentStream.countBitsSet(position - streamPosition);
+                streamPosition = position;
+            }
+
+            streamPosition++;
+            if (presentStream.nextBit()) {
+                // not null
+                if (nullsAllowed) {
+                    nulls[i] = false;
+                }
+                nonNullPositions[nonNullPositionCount] = nonNullPositionCount + nonNullSkipped;
+                nonNullPositionCount++;
+            }
+            else {
+                if (nullsAllowed) {
+                    nulls[i] = true;
+                }
+            }
+        }
+
+        readOffset = offset + streamPosition;
+
+        if (!nonNullsAllowed) {
+            outputPositionCount = positionCount - nonNullPositionCount;
+            allNulls = true;
+        }
+        else {
+            nestedLengths = ensureCapacity(nestedLengths, nonNullPositionCount);
+            Arrays.fill(nestedLengths, 0);
+
+            for (int keyIndex = 0; keyIndex < keyCount; keyIndex++) {
+                BooleanInputStream inMapStream = inMapStreams.get(keyIndex);
+
+                nestedPositions[keyIndex] = ensureCapacity(nestedPositions[keyIndex], nonNullPositionCount);
+                inMap[keyIndex] = ensureCapacity(inMap[keyIndex], nonNullPositionCount);
+
+                int nestedStreamPosition = 0;
+
+                int nestedSkipped = 0;
+                int nestedPositionCount = 0;
+                for (int i = 0; i < nonNullPositionCount; i++) {
+                    int position = nonNullPositions[i];
+                    if (position > nestedStreamPosition) {
+                        nestedSkipped += inMapStream.countBitsSet(position - nestedStreamPosition);
+                        nestedStreamPosition = position;
+                    }
+
+                    nestedStreamPosition++;
+                    if (inMapStream.nextBit()) {
+                        nestedPositions[keyIndex][nestedPositionCount] = nestedPositionCount + nestedSkipped;
+                        nestedPositionCount++;
+
+                        nestedLengths[i]++;
+                        inMap[keyIndex][i] = true;
+                    }
+                    else {
+                        inMap[keyIndex][i] = false;
+                    }
+                }
+
+                if (nonNullSkipped > nestedStreamPosition - nonNullPositionCount) {
+                    inMapStream.skip(nonNullSkipped - (nestedStreamPosition - nonNullPositionCount));
+                }
+
+                nestedPositionCounts[keyIndex] = nestedPositionCount;
+
+                if (nestedPositionCount > 0) {
```

Comment:
you lost me here. What's going on?

Reply:
nestedPositionCount might be zero if all requested rows are nulls or don't have values for this particular key